### ISSUE TITLE
Inconsistent font loading in Firefox

### ISSUE THREAD

**Issue by Drovolon** [foundryvtt/foundryvtt#11070]
### What happened?

*Sometimes*, I've noticed some fonts missing, most notably the `Pathfinder2eActions` system font - which means that action icons are displayed as literal numbers (e.g. `1` or `3`, most notably on the character sheet). I've spent about a couple days trying to debug this.

I've been rooting around in the font loading code in `resources/app/public/scripts/foundry.js`. Through judiciously adding log statements, I've found (at least) the following line in `loadFont()` *hangs* in Firefox sometimes:

```js
await fontFace.load();
```

When it does this, all(?) Foundry font loading stops and the top-level timeout in `_loadFonts` goes to maximum before `Promise.race()` gives up and the app continues onward. For the hung fonts, font loading never continues on its own; I assume there's some hung background thread, though I find it very hard to follow the JS async execution model.

Where font(s) hang seems inconsistent. I've seen as few as 3 fonts successfully load and as many as ~50+, and quite often (roughly ~8 in 10 times) all of them load just fine. But, when it breaks, fonts in the app are in a partially-loaded state, and no amount of waiting has them load AFAICT. When in this state, running `FontConfig._loadFonts()` in the console fixes it right up.

Weirdly, moving `fontFace.load()` into its own timeout-able function *seemingly* "fixes" the problem:

```js
  static loadFontFaceWithTimeout(fontFace, timeout = 3000) {
    return new Promise((resolve, reject) => {
      const timer = setTimeout(() => {
        reject(new Error('fontFace.load() timed out!'));
      }, timeout);
      fontFace.load().then(() => {
        clearTimeout(timer);
        resolve();
      }).catch(err => {
        clearTimeout(timer);
        reject(err);
      });
    });
  }

// ... later, in loadFont:

        await this.loadFontFaceWithTimeout(fontFace).catch(err => {
          console.error(`Failed to load FontFace for family "${family}", variant: `, fontVariant, 'Error:', err);
          throw err;
        });
```

As far as I know, this _shouldn't_ "fix" the problem. But, so it seemingly does... 

I've [attached log lines](https://github.com/foundryvtt/foundryvtt/files/15478184/console-export-2024-5-28_17-50-11.txt) of my heavily-modified local copy of foundry.js (I can copy/paste the code if you want, wasn't sure if that was kosher since I don't think it's open source?):

- "Fonts loading..." was inserted at the top of `_loadFonts`.
- "Enqueuing load" was right before `promises.push()` in `_loadFonts`.
- "Starting to load font family" was in the top of `loadFont`.
- "Loading font variant" was in the first for-loop in `loadFont`.
- "Created FontFace" is right after `_createFontFace`
- "Font face loaded" is right after `await fontFace.load()`
- "Font family loading" was before `documents.font.load` in `loadFont`.

Notice font loading starts at 17:49:23.097, then seemingly "hangs" at 17:49:24.876 (after loading Roboto) until 17:49:38.125 when the 15 second timeout is reached. (I increased the timeout from 4.5s to 15s.)

I wrote a little script to parse those logs:

```
# those put into Promises.all() but never reached the end of loadFont():
not_loaded=Counter({'MusiQwik': 1, 'Oriental': 1, 'Scrapbook Chinese': 1, 'Valmaric': 1})
----
# those who reached `await documents.font.load` but didn't reach the end of loadFont():
not_loaded_2=Counter()
----
# those who had _createFontFace succeed but didn't have `await fontFace.load()` finish:
face_not_loaded=Counter({'MusiQwik': 1, 'Oriental': 1, 'Scrapbook Chinese': 1, 'Valmaric': 1})
----
```

<details>
<summary>(script)</summary>

```python
import re
from dataclasses import dataclass, field
from collections import Counter

@dataclass
class LogLine:
    regex: re.Pattern
    matches: Counter[str] = field(default_factory=Counter)

log_types = {
    # did we put the promise in Promises.all()?
    'enqueue': LogLine(
        regex=re.compile(r'Enqueuing load for "([^"]+)"'),
    ),

    # when did we start loadFont?
    'load_started': LogLine(
        regex=re.compile(r'Starting to load font family "([^"]+)'),
    ),
    # what did the loadFont promise return?
    'load_result': LogLine(
        regex=re.compile(r'Load result for "([^"]+)": (\w+)'),
    ),

    # when did we invoke _createFontFace?
    'variant_created': LogLine(
        regex=re.compile(r'Loading font variant for family "([^"]+)"'),
    ),
    # when did _createFontFace complete?
    'face_created': LogLine(
        regex=re.compile(r'Created FontFace for family "([^"]+)"'),
    ),
    # when did fontFace.load() complete?
    'face_loaded': LogLine(
        regex=re.compile(r'Font face loaded for family "([^"]+)"'),
    ),

    # when did we start awaiting document.fonts.load?
    'font_loading': LogLine(
        regex=re.compile(r'Font family "([^"]+)" loading...'),
    ),
    # when did documents.font.load complete?
    'font_loaded': LogLine(
        regex=re.compile(r'Font family "([^"]+)" loaded.'),
    ),
}

with open('console-export-2024-5-28_17-50-11.txt', 'r') as logs:
    for line in logs:
        for log_type in log_types.values():
           if m := log_type.regex.search(line):
               log_type.matches[m.group(1)] += 1


print('# those put into Promises.all() but never reached the end of loadFont():')
not_loaded = log_types['enqueue'].matches - log_types['font_loaded'].matches
print(f'{not_loaded=}')
print('----')

print('# those who reached `await documents.font.load` but didn\'t reach the end of loadFont():')
not_loaded_2 = log_types['font_loading'].matches - log_types['font_loaded'].matches
print(f'{not_loaded_2=}')
print('----')

print('# those who had _createFontFace succeed but didn\'t have `await fontFace.load()` finish:')
face_not_loaded = log_types['face_created'].matches- log_types['face_loaded'].matches
print(f'{face_not_loaded=}')
print('----')
```

</details>

I'm sorry for such a weird bug report. I don't even know if it's reproducible on any other system. If you need more information, please let me know.

### What ways of accessing Foundry can you encounter this issue in?

- [ ] Native App (Electron)
- [ ] Chrome
- [X] Firefox
- [ ] Safari
- [ ] Other

### Reproduction Steps

I can only repro in Firefox (I tried Edge as well). Running Windows 11. Foundry is running on a e2-standard-2 GCE instance, behind Caddy as a reverse-proxy.

1. Use Pathfinder 2e module, version 6.0.0-beta2 (not sure if the version matters)
1. To make it easier to reproduce, use the Polyglot module. (I can repro without it, but less consistently.)
1. To make it easier to reproduce, disable caching in the network tab of dev tools.
1. Raise the `_loadFonts` timeout (in code) to something much higher, say 20 seconds. (This makes it more obvious when the condition triggers.)
1. With the console open, keep reloading the app until you notice the canvas takes ~20 seconds to appear. You've probably got missing fonts then. This may take like, a dozen retries.

### What core version are you reporting this for?

Version 12 Stable, Build 324

### Bug Checklist

- [ ] The issue occurs while all Modules are disabled

**Comment by WolfWings** [comment]
For anyone else stumbling into this bug, the following snippet as a 'bookmarklet' lets you fix the issue easily when (not if) it occurs at least as a sort of workaround:

`javascript:(function(){FontConfig._loadFonts();})();`

**Comment by Drovolon** [comment]
This seemingly still happens in Foundry 12 build 331.

Here is a patch file that can be more easily applied server-side (note: **only works for v12 build 331!**): [font-load-foundryjs-12.331.patch](https://github.com/user-attachments/files/16646774/font-load-foundryjs-12.331.patch)

For example, on a CLI

```sh
# assuming patch is placed at: /tmp/font-load-foundryjs-12.331.patch
$ cd /opt/foundryvtt/resources/app/public/scripts
$ cp foundry.js foundry.js.old   # make a backup!
$ patch --dry-run < /tmp/font-load-foundryjs-12.331.patch
checking foundry.js
$ patch < /tmp/font-load-foundryjs-12.331.patch
patching foundry.js
```

**Comment by aaclayton** [comment]
It looks like the font loading API supports event emitters for `loadingdone` and `loadingerror` events. I wonder if those events can be used rather than relying on an arbitrary timeout.
